File: go/inspect_test.go
package baseh

import (
	"math/big"
	"strings"
	"testing"
)

func mustCodec(t *testing.T, p Profile) *Codec {
	t.Helper()
	h, err := New(p)
	if err != nil {
		t.Fatalf("New(%s): %v", p.ProfileID, err)
	}
	return h
}

func TestInspectFixedEmpty(t *testing.T) {
	medium := mustCodec(t, MediumV1())
	for _, input := range []string{"", "   ", " - \t"} {
		if r := medium.Inspect(input); r.State != InspectEmpty {
			t.Errorf("Inspect(%q).State = %s, want %s", input, r.State, InspectEmpty)
		}
	}
}

func TestInspectFixedTypingPrefixes(t *testing.T) {
	medium := mustCodec(t, MediumV1())
	canonical, err := medium.Encode(big.NewInt(123456789))
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	raw := strings.ReplaceAll(canonical, "-", "")
	for n := 1; n < 8; n++ {
		r := medium.Inspect(raw[:n])
		if r.State != InspectTyping {
			t.Fatalf("prefix %d: state = %s, want %s", n, r.State, InspectTyping)
		}
		if got := strings.ReplaceAll(r.Typed, "-", ""); got != raw[:n] {
			t.Errorf("prefix %d: typed = %q, want symbols %q", n, r.Typed, raw[:n])
		}
		if want := float64(n) / 8; r.Progress != want {
			t.Errorf("prefix %d: progress = %v, want %v", n, r.Progress, want)
		}
	}
	// Separators inserted as far as the groups go (grouping [4, 4]).
	r := medium.Inspect(raw[:5])
	if r.State != InspectTyping || r.Typed != raw[:4]+"-"+raw[4:5] {
		t.Errorf("typed = %q (%s), want %q", r.Typed, r.State, raw[:4]+"-"+raw[4:5])
	}
}

func TestInspectFixedTypingNormalizesCaseAndAliases(t *testing.T) {
	medium := mustCodec(t, MediumV1())
	canonical, err := medium.Encode(big.NewInt(123456789))
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	raw := strings.ReplaceAll(canonical, "-", "")
	lower := medium.Inspect(strings.ToLower(raw[:5]))
	if lower.State != InspectTyping || lower.Typed != raw[:4]+"-"+raw[4:5] {
		t.Errorf("lowercase typed = %q (%s), want %q", lower.Typed, lower.State, raw[:4]+"-"+raw[4:5])
	}
	// Alias sources typed mid-code normalize to their targets (O -> 0 etc.).
	cloneProfile := MediumV1()
	cloneProfile.Permutation = Permutation{}
	cloneProfile.Profanity = Profanity{}
	cloneProfile.MaxRepetition = 0
	clone := mustCodec(t, cloneProfile)
	aliased := clone.Inspect("OIL")
	if aliased.State != InspectTyping || aliased.Typed != "011" {
		t.Errorf("aliased typed = %q (%s), want %q", aliased.Typed, aliased.State, "011")
	}
}

func TestInspectFixedTypingIgnoresWhitespaceAndSeparators(t *testing.T) {
	medium := mustCodec(t, MediumV1())
	canonical, err := medium.Encode(big.NewInt(123456789))
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	raw := strings.ReplaceAll(canonical, "-", "")
	messy := " " + raw[:2] + " -" + raw[2:5] + "\t"
	r := medium.Inspect(messy)
	if r.State != InspectTyping {
		t.Fatalf("state = %s, want %s", r.State, InspectTyping)
	}
	if got := strings.ReplaceAll(r.Typed, "-", ""); got != raw[:5] {
		t.Errorf("typed = %q, want symbols %q", r.Typed, raw[:5])
	}
}

func TestInspectFixedPaddedPrefixStillTyping(t *testing.T) {
	// Spec 3.4: find a short input whose re-padded form validates (the
	// cookbook's "false green"), on a filter-free clone so the scan is not
	// disturbed by the blocklist or repetition filter. inspect must still
	// report typing for it.
	cloneProfile := MediumV1()
	cloneProfile.Profanity = Profanity{}
	cloneProfile.MaxRepetition = 0
	clone := mustCodec(t, cloneProfile)
	medium := mustCodec(t, MediumV1())
	found := ""
	for id := int64(0); id < 200000 && found == ""; id++ {
		code, err := clone.Encode(big.NewInt(id))
		if err != nil {
			t.Fatalf("Encode(%d): %v", id, err)
		}
		raw := strings.ReplaceAll(code, "-", "")
		stripped := strings.TrimLeft(raw, "0")
		if stripped == "" {
			stripped = raw[len(raw)-1:]
		}
		if len(stripped) < len(raw) && len(stripped) >= 2 && clone.Validate(stripped, nil).Valid {
			found = stripped
		}
	}
	if found == "" {
		t.Fatal("no false-green prefix found in scan window")
	}
	if r := medium.Inspect(found); r.State != InspectTyping {
		t.Errorf("Inspect(%q).State = %s, want %s", found, r.State, InspectTyping)
	}
}

func TestInspectFixedValid(t *testing.T) {
	medium := mustCodec(t, MediumV1())
	canonical, err := medium.Encode(big.NewInt(123456789))
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	r := medium.Inspect(canonical)
	if r.State != InspectValid || r.ID.Cmp(big.NewInt(123456789)) != 0 || r.CanonicalCode != canonical {
		t.Fatalf("Inspect(%q) = %+v, want valid id 123456789 code %q", canonical, r, canonical)
	}
	// No separators, lowercase, surrounding whitespace all reach valid.
	messy := " " + strings.ToLower(strings.ReplaceAll(canonical, "-", "")) + " "
	r2 := medium.Inspect(messy)
	if r2.State != InspectValid || r2.ID.Cmp(r.ID) != 0 || r2.CanonicalCode != r.CanonicalCode {
		t.Errorf("Inspect(%q) = %+v, want %+v", messy, r2, r)
	}
}

func TestInspectFixedAliasTypedCompleteCode(t *testing.T) {
	cloneProfile := MediumV1()
	cloneProfile.Profanity = Profanity{}
	cloneProfile.MaxRepetition = 0
	clone := mustCodec(t, cloneProfile)
	for id := int64(1); id < 100000; id++ {
		code, err := clone.Encode(big.NewInt(id))
		if err != nil {
			t.Fatalf("Encode(%d): %v", id, err)
		}
		raw := strings.ReplaceAll(code, "-", "")
		if strings.Contains(raw, "8") {
			r := clone.Inspect(strings.Replace(raw, "8", "B", 1)) // B -> 8
			if r.State != InspectValid || r.ID.Cmp(big.NewInt(id)) != 0 {
				t.Fatalf("Inspect(B-typed %q) = %+v, want valid id %d", raw, r, id)
			}
			return
		}
	}
	t.Fatal("no code containing 8 found")
}

func TestInspectFixedInvalidChecksum(t *testing.T) {
	medium := mustCodec(t, MediumV1())
	canonical, err := medium.Encode(big.NewInt(77))
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	raw := []byte(strings.ReplaceAll(canonical, "-", ""))
	if raw[6] == '2' {
		raw[6] = '3'
	} else {
		raw[6] = '2'
	}
	r := medium.Inspect(string(raw))
	if r.State != InspectInvalid || r.Reason != INVALID_CHECKSUM {
		t.Errorf("Inspect(%q) = %+v, want invalid %s", raw, r, INVALID_CHECKSUM)
	}
}

func TestInspectFixedBadChar(t *testing.T) {
	medium := mustCodec(t, MediumV1())
	for _, input := range []string{"12@", "1234-56@8"} {
		if r := medium.Inspect(input); r.State != InspectBadChar {
			t.Errorf("Inspect(%q).State = %s, want %s", input, r.State, InspectBadChar)
		}
	}
}

func TestInspectChecksumOnlySymbolInBodyIsInvalid(t *testing.T) {
	// U is in the Heavy checksum alphabet but not its body alphabet: it
	// passes the union-membership gate and fails under validate, exactly
	// like the shared error vector (heavy "U00000A" -> INVALID_CHARACTER).
	heavy := mustCodec(t, HeavyV1())
	r := heavy.Inspect("U000000A")
	if r.State != InspectInvalid || r.Reason != INVALID_CHARACTER {
		t.Errorf("Inspect = %+v, want invalid %s", r, INVALID_CHARACTER)
	}
}

func TestInspectFixedTooLong(t *testing.T) {
	medium := mustCodec(t, MediumV1())
	for _, input := range []string{"00000000C", "0000-0000-C"} {
		if r := medium.Inspect(input); r.State != InspectTooLong {
			t.Errorf("Inspect(%q).State = %s, want %s", input, r.State, InspectTooLong)
		}
	}
}

func TestInspectFixedNoChecksumProfile(t *testing.T) {
	minimum := mustCodec(t, MinimumV1())
	canonical, err := minimum.Encode(big.NewInt(42))
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	if r := minimum.Inspect(canonical); r.State != InspectValid {
		t.Errorf("Inspect(%q).State = %s, want %s", canonical, r.State, InspectValid)
	}
	if r := minimum.Inspect(canonical[:3]); r.State != InspectTyping {
		t.Errorf("prefix state = %s, want %s", r.State, InspectTyping)
	}
}

func TestInspectExpandableEmptyAndTyping(t *testing.T) {
	expandable := mustCodec(t, ExpandableV1())
	if r := expandable.Inspect(""); r.State != InspectEmpty {
		t.Errorf("empty: state = %s", r.State)
	}
	cases := []struct {
		input    string
		typed    string
		progress float64
	}{
		{"1", "1", 0.25},
		{"12", "12", 0.5},
		{"123", "123", 0.75},
		{"ab", "AB", 0.5}, // below separatorMinLength the typing render is bare
		{"O", "0", 0.25},  // alias O -> 0; 0 is a checksum-alphabet symbol
	}
	for _, c := range cases {
		r := expandable.Inspect(c.input)
		if r.State != InspectTyping || r.Typed != c.typed || r.Progress != c.progress {
			t.Errorf("Inspect(%q) = %+v, want typing %q progress %v", c.input, r, c.typed, c.progress)
		}
	}
}

func TestInspectExpandableGenerationBoundaries(t *testing.T) {
	expandable := mustCodec(t, ExpandableV1())
	for _, id := range []int64{0, 39304, 1375640} {
		code, err := expandable.Encode(big.NewInt(id))
		if err != nil {
			t.Fatalf("Encode(%d): %v", id, err)
		}
		r := expandable.Inspect(code)
		if r.State != InspectValid || r.ID.Cmp(big.NewInt(id)) != 0 || r.CanonicalCode != code {
			t.Errorf("Inspect(%q) = %+v, want valid id %d code %q", code, r, id, code)
		}
	}
}

func TestInspectExpandableCompleteLengthsAreNeverTyping(t *testing.T) {
	expandable := mustCodec(t, ExpandableV1())
	code, err := expandable.Encode(big.NewInt(777)) // generation 4
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	raw := strings.ReplaceAll(code, "-", "")
	five := raw + "A" // wrong-length presentation, checksum fails (spec 19.7)
	r := expandable.Inspect(five)
	if r.State != InspectInvalid || r.Reason != INVALID_CHECKSUM {
		t.Errorf("Inspect(%q) = %+v, want invalid %s", five, r, INVALID_CHECKSUM)
	}
}

func TestInspectExpandableZeroInBodyPosition(t *testing.T) {
	expandable := mustCodec(t, ExpandableV1())
	code, err := expandable.Encode(big.NewInt(777))
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	raw := strings.ReplaceAll(code, "-", "")
	for _, bad := range []string{"0" + raw[1:], "O" + raw[1:]} {
		r := expandable.Inspect(bad)
		if r.State != InspectInvalid || r.Reason != INVALID_CHARACTER {
			t.Errorf("Inspect(%q) = %+v, want invalid %s", bad, r, INVALID_CHARACTER)
		}
	}
}

func TestInspectExpandableBadCharAndTooLong(t *testing.T) {
	expandable := mustCodec(t, ExpandableV1())
	for _, input := range []string{"A@", "ABCD@"} {
		if r := expandable.Inspect(input); r.State != InspectBadChar {
			t.Errorf("Inspect(%q).State = %s, want %s", input, r.State, InspectBadChar)
		}
	}
	if r := expandable.Inspect(strings.Repeat("A", 33)); r.State != InspectTooLong {
		t.Errorf("33 symbols: state = %s, want %s", r.State, InspectTooLong)
	}
	// 32 real symbols pass the length gate and land on validate.
	if r := expandable.Inspect(strings.Repeat("A", 32)); r.State != InspectInvalid {
		t.Errorf("32 symbols: state = %s, want %s", r.State, InspectInvalid)
	}
}

func TestInspectExpandableMessyCompleteCode(t *testing.T) {
	expandable := mustCodec(t, ExpandableV1())
	code, err := expandable.Encode(big.NewInt(1375640))
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	raw := strings.ReplaceAll(code, "-", "")
	r := expandable.Inspect(" " + raw[:3] + " - " + raw[3:])
	if r.State != InspectValid || r.ID.Cmp(big.NewInt(1375640)) != 0 || r.CanonicalCode != code {
		t.Errorf("Inspect = %+v, want valid id 1375640 code %q", r, code)
	}
}

func TestInspectFacadeMatchesInstance(t *testing.T) {
	expandable := mustCodec(t, ExpandableV1())
	code, err := expandable.Encode(big.NewInt(42))
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	for _, input := range []string{"", "1", "AB@", strings.Repeat("A", 33), code} {
		facade := Inspect(input)
		instance := expandable.Inspect(input)
		if facade.State != instance.State ||
			facade.Typed != instance.Typed ||
			facade.Progress != instance.Progress ||
			facade.Reason != instance.Reason ||
			facade.CanonicalCode != instance.CanonicalCode ||
			(facade.ID == nil) != (instance.ID == nil) ||
			(facade.ID != nil && facade.ID.Cmp(instance.ID) != 0) {
			t.Errorf("Inspect(%q): facade %+v != instance %+v", input, facade, instance)
		}
	}
	if r := Inspect(code); r.State != InspectValid {
		t.Errorf("facade Inspect(%q).State = %s, want %s", code, r.State, InspectValid)
	}
}

File: js/test/inspect.test.ts
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  Baseh,
  basehMediumV1,
  basehMinimumV1,
  basehHeavyV1,
  basehExpandableV1,
  inspect,
  type InspectResult
} from "../src/index.js";

const medium = new Baseh(basehMediumV1()); // fixed, expected 8, grouping [4,4]
const expandable = new Baseh(basehExpandableV1()); // minLength 4, separatorMinLength 6

function statesOf(inputs: string[]): InspectResult[] {
  return inputs.map((i) => medium.inspect(i));
}

describe("inspect: fixed mode (baseh-medium-v1)", () => {
  it("empty states", () => {
    assert.deepEqual(medium.inspect(""), { state: "empty" });
    assert.deepEqual(medium.inspect("   "), { state: "empty" });
    assert.deepEqual(medium.inspect(" - \t"), { state: "empty" });
  });

  it("typing: prefixes of a real code carry normalized symbols and progress", () => {
    const canonical = medium.encode(123456789n); // e.g. "XXXX-XXXX"
    const raw = canonical.replaceAll("-", "");
    for (let n = 1; n < 8; n += 1) {
      const r = medium.inspect(raw.slice(0, n));
      assert.equal(r.state, "typing", `prefix ${n}`);
      if (r.state !== "typing") throw new Error("unreachable");
      assert.equal(r.typed.replaceAll("-", ""), raw.slice(0, n));
      assert.equal(r.progress, n / 8);
    }
    // separators inserted as far as the groups go (grouping [4, 4])
    const r5 = medium.inspect(raw.slice(0, 5));
    assert.equal(r5.state, "typing");
    if (r5.state === "typing") assert.equal(r5.typed, raw.slice(0, 4) + "-" + raw.slice(4, 5));
  });

  it("typing: lowercase and aliases normalize while typing", () => {
    const canonical = medium.encode(123456789n);
    const raw = canonical.replaceAll("-", "");
    const lower = medium.inspect(raw.slice(0, 5).toLowerCase());
    assert.equal(lower.state, "typing");
    if (lower.state === "typing") assert.equal(lower.typed, raw.slice(0, 4) + "-" + raw[4]);
    // alias source typed mid-code normalizes to its target (O -> 0 etc.)
    const aliasProfile = new Baseh({ ...basehMediumV1(), permutation: { enabled: false }, profanity: { mode: "none" }, maxRepetition: 0 });
    const aliased = aliasProfile.inspect("OIL");
    assert.equal(aliased.state, "typing");
    if (aliased.state === "typing") assert.equal(aliased.typed, "011");
  });

  it("typing: whitespace and stray separators are ignored for counting", () => {
    const canonical = medium.encode(123456789n);
    const raw = canonical.replaceAll("-", "");
    const messy = " " + raw.slice(0, 2) + " -" + raw.slice(2, 5) + "\t";
    const r = medium.inspect(messy);
    assert.equal(r.state, "typing");
    if (r.state === "typing") assert.equal(r.typed.replaceAll("-", ""), raw.slice(0, 5));
  });

  it("spec 3.4: a padded prefix that passes the checksum still reports typing", () => {
    // Find a short input whose re-padded form validates (the cookbook's
    // "false green"), on a filter-free clone so the scan is not disturbed by
    // the blocklist or repetition filter.
    const clone = new Baseh({ ...basehMediumV1(), profanity: { mode: "none" }, maxRepetition: 0 });
    let found: string | null = null;
    for (let id = 0n; id < 200000n && found === null; id += 1n) {
      const raw = clone.encode(id).replaceAll("-", "");
      const stripped = raw.replace(/^0+(?=.)/, "");
      if (stripped.length < raw.length && stripped.length >= 2 && clone.validate(stripped).valid) {
        found = stripped;
      }
    }
    assert.notEqual(found, null, "no false-green prefix found in scan window");
    const r = medium.inspect(found as string);
    assert.equal(r.state, "typing");
  });

  it("valid: complete code, with id and canonicalCode", () => {
    const canonical = medium.encode(123456789n);
    const r = medium.inspect(canonical);
    assert.deepEqual(r, { state: "valid", id: 123456789n, canonicalCode: canonical });
    // no separators, lowercase, surrounding whitespace all reach valid
    assert.deepEqual(medium.inspect(" " + canonical.replaceAll("-", "").toLowerCase() + " "), r);
  });

  it("valid: alias-typed complete code decodes", () => {
    const clone = new Baseh({ ...basehMediumV1(), profanity: { mode: "none" }, maxRepetition: 0 });
    // find a code containing 8, type it with B (B -> 8)
    for (let id = 1n; id < 100000n; id += 1n) {
      const raw = clone.encode(id).replaceAll("-", "");
      if (raw.includes("8")) {
        const r = clone.inspect(raw.replace("8", "B"));
        assert.equal(r.state, "valid");
        if (r.state === "valid") assert.equal(r.id, id);
        return;
      }
    }
    throw new Error("no code containing 8 found");
  });

  it("invalid: complete code with a wrong checksum carries the reason", () => {
    const canonical = medium.encode(77n);
    const raw = canonical.replaceAll("-", "");
    const badCheck = raw[6] === "2" ? "3" : "2";
    const bad = raw.slice(0, 6) + badCheck + raw[7];
    const r = medium.inspect(bad);
    assert.deepEqual(r, { state: "invalid", reason: "INVALID_CHECKSUM" });
  });

  it("bad-char: symbol outside both alphabets, typing or complete", () => {
    assert.deepEqual(medium.inspect("12@"), { state: "bad-char" });
    assert.deepEqual(medium.inspect("1234-56@8"), { state: "bad-char" });
  });

  it("a checksum-only symbol in the body region is invalid, not bad-char", () => {
    // U is in the Heavy checksum alphabet but not its body alphabet: it
    // passes the union-membership gate and fails under validate, exactly
    // like the shared error vector (heavy "U00000A" -> INVALID_CHARACTER).
    const heavy = new Baseh(basehHeavyV1());
    const r = heavy.inspect("U000000A");
    assert.deepEqual(r, { state: "invalid", reason: "INVALID_CHARACTER" });
  });

  it("too-long: more than bodyLength + checksumLength symbols", () => {
    assert.deepEqual(medium.inspect("00000000C"), { state: "too-long" });
    assert.deepEqual(medium.inspect("0000-0000-C"), { state: "too-long" });
  });

  it("no-checksum fixed profile: every complete length validates", () => {
    const minimum = new Baseh(basehMinimumV1()); // 6 symbols, no checksum
    const canonical = minimum.encode(42n);
    const r = minimum.inspect(canonical);
    assert.equal(r.state, "valid");
    for (const s of statesOf([canonical.slice(0, 3)])) assert.equal(s.state, "typing");
  });
});

describe("inspect: expandable mode (baseh-expandable-v1)", () => {
  it("empty and below-minLength typing", () => {
    assert.deepEqual(expandable.inspect(""), { state: "empty" });
    assert.deepEqual(expandable.inspect("1"), { state: "typing", typed: "1", progress: 0.25 });
    assert.deepEqual(expandable.inspect("12"), { state: "typing", typed: "12", progress: 0.5 });
    assert.deepEqual(expandable.inspect("123"), { state: "typing", typed: "123", progress: 0.75 });
    // below separatorMinLength the typing render is bare (B aliases to 8)
    assert.deepEqual(expandable.inspect("ab"), { state: "typing", typed: "A8", progress: 0.5 });
    // aliases normalize while typing (O -> 0, and 0 is a checksum-alphabet symbol)
    assert.deepEqual(expandable.inspect("O"), { state: "typing", typed: "0", progress: 0.25 });
  });

  it("generation boundaries: minLength is the first complete length", () => {
    const code4 = expandable.encode(0n); // first id, generation 4
    assert.equal(code4.length, 4);
    assert.deepEqual(expandable.inspect(code4), { state: "valid", id: 0n, canonicalCode: code4 });
    const code5 = expandable.encode(19683n); // first id of generation 5
    assert.equal(code5.length, 5);
    assert.deepEqual(expandable.inspect(code5), { state: "valid", id: 19683n, canonicalCode: code5 });
    const code6 = expandable.encode(551124n); // first id of generation 6, renders with a hyphen
    assert.equal(code6.length, 7);
    assert.deepEqual(expandable.inspect(code6), { state: "valid", id: 551124n, canonicalCode: code6 });
  });

  it("every length >= minLength is complete: a bad checksum is invalid, not typing", () => {
    // 5 symbols that fail the generation-5 checksum
    const sample = expandable.encode(777n).replaceAll("-", ""); // generation 4
    const five = sample + "A"; // wrong-length presentation, checksum fails (spec 19.7)
    const r = expandable.inspect(five);
    assert.equal(r.state, "invalid");
    if (r.state === "invalid") assert.equal(r.reason, "INVALID_CHECKSUM");
  });

  it("0 or O in a body position is invalid with INVALID_CHARACTER", () => {
    const sample = expandable.encode(777n).replaceAll("-", "");
    for (const bad of ["0" + sample.slice(1), "O" + sample.slice(1)]) {
      const r = expandable.inspect(bad);
      assert.deepEqual(r, { state: "invalid", reason: "INVALID_CHARACTER" });
    }
  });

  it("bad-char and too-long", () => {
    assert.deepEqual(expandable.inspect("A@"), { state: "bad-char" });
    assert.deepEqual(expandable.inspect("ABCD@"), { state: "bad-char" });
    assert.deepEqual(expandable.inspect("A".repeat(33)), { state: "too-long" });
    // 32 real symbols pass the length gate and land on validate
    const r = expandable.inspect("A".repeat(32));
    assert.equal(r.state, "invalid");
  });

  it("whitespace and separators in a complete code still reach valid", () => {
    const code6 = expandable.encode(551124n);
    const raw = code6.replaceAll("-", "");
    const r = expandable.inspect(" " + raw.slice(0, 3) + " - " + raw.slice(3));
    assert.deepEqual(r, { state: "valid", id: 551124n, canonicalCode: code6 });
  });
});

describe("inspect: zero-config facade", () => {
  it("matches a default-profile instance", () => {
    for (const input of ["", "1", "AB@", "A".repeat(33), expandable.encode(42n)]) {
      assert.deepEqual(inspect(input), expandable.inspect(input));
    }
    const r = inspect(expandable.encode(42n));
    assert.equal(r.state, "valid");
  });
});
